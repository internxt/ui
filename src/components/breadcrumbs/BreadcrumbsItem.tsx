import { DropTargetMonitor, useDrop } from 'react-dnd';
import { Dispatch } from 'redux';
import { FunctionComponent, SVGProps } from 'react';

export interface BreadcrumbItemData {
  uuid: string;
  label: string;
  icon: JSX.Element | null;
  active: boolean;
  isFirstPath?: boolean;
  dialog?: boolean;
  isBackup?: boolean;
  onClick?: () => void;
}

export interface BreadcrumbsMenuProps {
  item: BreadcrumbItemData;
  items: BreadcrumbItemData[];
  onItemClicked: (item: BreadcrumbItemData) => void;
}

/**
 * BreadcrumbsItem component
 *
 * @property {BreadcrumbItemData} item
 * - Data representing a single breadcrumb item, including label, icon, and other properties.
 *
 * @property {number} totalBreadcrumbsLength
 * - The total number of breadcrumb items, used for styling the first item and determining when to apply certain styles.
 *
 * @property {boolean} [isHiddenInList]
 * - If true, the breadcrumb is hidden in the list and only shown in a dropdown menu.
 *
 * @property {BreadcrumbItemData[]} items
 * - Array of all breadcrumb items, used for rendering all the breadcrumbs and their separators.
 *
 * @property {string} [breadcrumbButtonDataCy]
 * - Custom `data-cy` attribute applied to the breadcrumb button element.
 *
 * @property {FunctionComponent<BreadcrumbsMenuProps>} [menu]
 * - A custom menu component that can be shown for the breadcrumb item when it's not active or dialog-based.
 *
 * @property {Object[]} namePath
 * - Array of objects representing the path and UUID of the breadcrumb item.
 *
 * @property {boolean} isSomeItemSelected
 * - If true, indicates that at least one breadcrumb item is selected, affecting styling or behavior.
 *
 * @property {any[]} selectedItems
 * - Array of selected breadcrumb items, used to manage selected states and actions.
 *
 * @property {Function} onItemDropped
 * - Callback function that is triggered when a breadcrumb item is dropped.
 *
 * @property {Function} canItemDrop
 * - Determines if a breadcrumb item can be dropped. Used for validation during drag-and-drop operations.
 *
 * @property {FunctionComponent<SVGProps<SVGSVGElement>>} [itemComponent]
 * - Optional custom component for rendering an icon or other visual elements within the breadcrumb item.
 *
 * @property {string[]} acceptedTypes
 * - List of accepted drag-and-drop types for the breadcrumb item.
 *
 * @property {Dispatch} dispatch
 * - The Redux dispatch function for dispatching actions related to the breadcrumb item.
 */

export interface BreadcrumbsItemProps<T extends Dispatch> {
  item: BreadcrumbItemData;
  totalBreadcrumbsLength: number;
  isHiddenInList?: boolean;
  items: BreadcrumbItemData[];
  breadcrumbButtonDataCy?: string;
  menu?: (props: BreadcrumbsMenuProps) => JSX.Element;
  namePath: {
    name: string;
    uuid: string;
  }[];
  isSomeItemSelected: boolean;
  selectedItems: [];
  onItemDropped: (
    item: BreadcrumbItemData,
    namePath: {
      name: string;
      uuid: string;
    }[],
    isSomeItemSelected: boolean,
    selectedItems: [],
    dispatch: T,
  ) => (draggedItem: unknown, monitor: DropTargetMonitor) => Promise<void>;
  canItemDrop: (
    item: BreadcrumbItemData,
  ) => (draggedItem: unknown, monitor: DropTargetMonitor<unknown, unknown>) => boolean;
  itemComponent?: FunctionComponent<SVGProps<SVGSVGElement>>;
  acceptedTypes: string[];
  dispatch: T;
}

const BreadcrumbsItem = <T extends Dispatch>(props: BreadcrumbsItemProps<T>): JSX.Element => {
  const [{ isOver, canDrop }, drop] = useDrop(
    () => ({
      accept: props.acceptedTypes,
      collect: (monitor) => ({
        isOver: monitor.isOver(),
        canDrop: monitor.canDrop(),
      }),
      canDrop: props.canItemDrop(props.item),
      drop: props.onItemDropped(
        props.item,
        props.namePath,
        props.isSomeItemSelected,
        props.selectedItems,
        props.dispatch,
      ),
    }),
    [props.selectedItems],
  );

  const onItemClicked = (item: BreadcrumbItemData): void => {
    if (item.active) {
      item.onClick && item.onClick();
    }
  };
  const isDraggingOverClassNames = isOver && canDrop ? 'drag-over-effect' : '';

  return (
    <>
      {!props.item.active && !props.item.dialog && props.menu ? (
        <props.menu item={props.item} items={props.items} onItemClicked={onItemClicked} />
      ) : (
        <div
          ref={drop}
          className={`flex ${props.isHiddenInList ? 'w-full' : 'max-w-fit'} ${
            props.item.isFirstPath ? 'shrink-0 pr-1' : 'min-w-breadcrumb flex-1 px-1.5 py-1.5'
          } cursor-pointer flex-row items-center truncate font-medium ${isDraggingOverClassNames}
        ${
          !props.item.active || (props.item.isFirstPath && props.totalBreadcrumbsLength === 1)
            ? 'text-gray-80'
            : 'text-gray-50 hover:text-gray-80'
        }`}
          key={props.item.uuid}
          onClick={() => onItemClicked(props.item)}
          onKeyDown={() => {}}
          data-cy={props?.breadcrumbButtonDataCy}
        >
          {props.itemComponent && <props.itemComponent className="h-5 w-5" />}
          {props.item.icon ? props.item.icon : null}
          {props.item.label ? (
            <span
              className={`max-w-sm flex-1 cursor-pointer truncate ${props.isHiddenInList && 'pl-3 text-base'}`}
              title={props.item.label}
            >
              {props.item.label}
            </span>
          ) : null}
        </div>
      )}
    </>
  );
};

export default BreadcrumbsItem;
