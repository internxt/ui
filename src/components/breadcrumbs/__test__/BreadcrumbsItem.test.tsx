import React from 'react';
import { afterEach, describe, it, vi, expect, beforeEach } from 'vitest';
import { render, fireEvent } from '@testing-library/react';
import BreadcrumbsItem, { BreadcrumbItemData, BreadcrumbsItemProps } from '../BreadcrumbsItem';
import { useDrop } from 'react-dnd';

vi.mock('react-dnd', () => ({
  useDrop: vi.fn(),
}));

describe('BreadcrumbsItem Component', () => {
  const mockedUseDrop = useDrop as unknown as ReturnType<typeof vi.fn>;
  const onItemDroppedMock = vi.fn();
  const onClickMock = vi.fn();
  const dispatchMock = vi.fn();

  const breadcrumbItem: BreadcrumbItemData = {
    uuid: '1',
    label: 'Item 1',
    icon: null,
    active: true,
    isFirstPath: false,
    dialog: false,
    onClick: onClickMock,
  };

  const defaultProps: BreadcrumbsItemProps<typeof dispatchMock> = {
    item: breadcrumbItem,
    totalBreadcrumbsLength: 3,
    items: [breadcrumbItem],
    namePath: [{ name: 'Root', uuid: 'root' }],
    isSomeItemSelected: false,
    selectedItems: [],
    canItemDrop: vi.fn(() => () => true),
    onItemDropped: onItemDroppedMock,
    dispatch: dispatchMock,
    acceptedTypes: ['ITEM_TYPE'],
  };

  const renderBreadcrumbsItem = (props = {}) => render(<BreadcrumbsItem {...defaultProps} {...props} />);

  beforeEach(() => {
    mockedUseDrop.mockImplementation(() => [{ isOver: false, canDrop: false }, vi.fn()]);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should match snapshot', () => {
    const breadcrumbsItem = renderBreadcrumbsItem();
    expect(breadcrumbsItem).toMatchSnapshot();
  });

  it('should render the component with correct data', () => {
    const { getByText } = renderBreadcrumbsItem();
    expect(getByText('Item 1')).toBeInTheDocument();
  });

  it('should call `onClick` when an active item is clicked', () => {
    const { getByText } = renderBreadcrumbsItem();
    const item = getByText('Item 1');
    fireEvent.click(item);
    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  it('should not call `onClick` if the item is not active', () => {
    const inactiveProps = {
      ...defaultProps,
      item: { ...breadcrumbItem, active: false },
    };
    const { getByText } = renderBreadcrumbsItem(inactiveProps);
    const item = getByText('Item 1');
    fireEvent.click(item);
    expect(onClickMock).not.toHaveBeenCalled();
  });

  it('should apply the `drag-over-effect` class when `isOver` and `canDrop` are true', () => {
    mockedUseDrop.mockImplementation(() => [{ isOver: true, canDrop: true }, vi.fn()]);
    const { getByText } = renderBreadcrumbsItem();
    const item = getByText('Item 1');
    expect(item.parentElement).toHaveClass('drag-over-effect');
  });

  it('should render the menu if the item is neither active nor a dialog', () => {
    const menuMock = vi.fn(() => <div>Menu</div>);
    const menuProps = { ...defaultProps, menu: menuMock, item: { ...breadcrumbItem, active: false, dialog: false } };

    const { getByText } = renderBreadcrumbsItem(menuProps);
    expect(getByText('Menu')).toBeInTheDocument();
  });

  it('should handle `drop` correctly', async () => {
    const dropMock = vi.fn(() => Promise.resolve());
    mockedUseDrop.mockImplementation(() => [{ isOver: false, canDrop: true }, dropMock]);

    const { getByText } = renderBreadcrumbsItem();

    global.DragEvent = class extends Event {
      dataTransfer: DataTransfer | null = null;

      constructor(type: string, eventInitDict?: DragEventInit) {
        super(type, eventInitDict);
        this.dataTransfer = eventInitDict?.dataTransfer || null;
      }
    } as unknown as typeof DragEvent;

    const dropArea = getByText('Item 1').parentElement;
    if (dropArea && 'dispatchEvent' in dropArea) {
      const dropEvent = new DragEvent('drop', { bubbles: true });
      dropArea.dispatchEvent(dropEvent);
    }

    expect(dropMock).toHaveBeenCalledTimes(1);
  });

  it('should render an icon if provided', () => {
    const itemWithIcon = {
      ...breadcrumbItem,
      icon: <svg data-testid="icon" />,
    };
    const { getByTestId } = renderBreadcrumbsItem({ item: { ...itemWithIcon } });
    expect(getByTestId('icon')).toBeInTheDocument();
  });

  it('should render `itemComponent` if provided', () => {
    const ItemComponentMock = (props: React.SVGProps<SVGSVGElement>) => <svg {...props} data-testid="custom-icon" />;
    const { getByTestId } = renderBreadcrumbsItem({ itemComponent: ItemComponentMock });
    expect(getByTestId('custom-icon')).toBeInTheDocument();
  });
});
