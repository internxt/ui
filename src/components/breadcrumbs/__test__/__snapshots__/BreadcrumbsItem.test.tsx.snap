// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BreadcrumbsItem Component > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex max-w-fit min-w-breadcrumb flex-1 px-1.5 py-1.5 cursor-pointer flex-row items-center truncate font-medium 
        text-gray-50 hover:text-gray-80"
      >
        <span
          class="max-w-sm flex-1 cursor-pointer truncate undefined"
          title="Item 1"
        >
          Item 1
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex max-w-fit min-w-breadcrumb flex-1 px-1.5 py-1.5 cursor-pointer flex-row items-center truncate font-medium 
        text-gray-50 hover:text-gray-80"
    >
      <span
        class="max-w-sm flex-1 cursor-pointer truncate undefined"
        title="Item 1"
      >
        Item 1
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
